packages/atclient/src/atclient_utils.c

```diff
@@ -0,0 +1,103 @@
+#include "atclient/atclient_utils.h"
+#include "atclient/atkeys.h"
+#include "atclient/connection.h"
+#include "atclient/stringutils.h"
+#include <atlogger/atlogger.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <pwd.h>
+#include <unistd.h>
+#include <string.h>
+
+#define TAG "atclient_utils"
+
+int atclient_utils_find_atserver_address(const char *atdirectory_host, const int atdirectory_port, const char *atsign,
+                                   char **atserver_host, int *atserver_port) {
```

Comment:
this need not be a char**, char* is already an address

Reply:
The function allocates and sets the pointer for you,

So it would be used like this:

```c

char *atserver_host = NULL;
int atserver_port = -1;

atclient_utils_find_atserver_address(/* ... */, &atserver_host, &atserver_port);

```

Reply:
If I were to pass just a `char *`, I would have to trust that the caller allocated sufficient space.

Instead, this function allocates just enough space for you so that no memory is wasted.

Example, if the `atserver_host` holds something like `www.google.com`, the caller could either allocate something like 512 bytes to ensure that there is *always* enough space, OR the function could dynamically allocate the space needed (in this case 14 bytes).